packages/core/src/sagas/incomingActivitySaga.js

```diff
@@ -20,38 +22,45 @@ function last(array, predicate) {
 }
 
 export default function* () {
-  yield whileConnected(function* (directLine, userID) {
-    yield observeEach(directLine.activity$, function* (activity) {
-      activity = { ...activity };
-
-      // Patch activity.from.role to make sure its either "bot", "user", or "channel"
-      if (!activity.from) {
-        activity.from = { role: 'channel' };
-      } else if (!activity.from.role) {
-        if (activity.from.id === userID) {
-          activity.from.role = 'user';
-        } else if (activity.from.id) {
-          activity.from.role = 'bot';
-        } else {
-          activity.from.role = 'channel';
-        }
-      }
-
-      yield put(incomingActivity(activity));
-
-      // Update suggested actions
-      const activities = yield select(({ activities }) => activities);
-      const lastMessageActivity = last(activities, ({ type }) => type === 'message');
-
-      if (
-        lastMessageActivity
-        && lastMessageActivity.from
-        && lastMessageActivity.from.role === 'bot'
-      ) {
-        const { suggestedActions: { actions } = {} } = lastMessageActivity;
-
-        yield put(setSuggestedActions(actions));
-      }
-    });
+  yield whileConnected(observeActivity);
+}
+
+function* observeActivity(directLine, userID) {
+  yield observeEach(directLine.activity$, function* (activity) {
+    activity = patchActivityWithFromRole(activity, userID);
+
+    yield put(incomingActivity(activity));
+
+    // Update suggested actions
+    const { activities } = yield select();
```

Comment:
I will disagree and commit here but it's odd that the convention is to pull _all_ state into scope and then parse it down when redux-saga provides a mechanism to document what part of the state it intended to be inspected via a selector. I see selectors and definitions working together, not mutually exclusive.

Reply:
I am unsure if the selector inside `select()` function will do any perf benefit if we scope it down, that's why I am changing it back to the bare style. Syntax-wise, it looks more readable and compact,

```js
// Bare approach
const { activities } = yield select();
```

Compare to the original selector function,

```js
// Inline selector
const activities = yield select(({ activities }) => activities);
```

Yes, if we have the selector, then we can simplify it as (shortest after uglify):

```js
// Effect
const activities = yield selectActivitiesFromStore();
```

Or
```js
// Selector factory
const activities = yield select(activitiesSelector());
```

> I prefer selector factory here because we can add additional selectors like, `activitiesSelector(({ type }) => type === 'message')` or `activitiesSelector(speakableActivity)`.

What do you think about the 4th approach (selector factory)? The implementation would looks like,

```js
export default function (predicate) {
  return ({ activities }) => predicate ? activities.filter(predicate) : activities;
}
```

But when we need to combine multiple selectors, it will start look a bit cumbersome, or we need to do 2 calls to `select()`,

```js
// With combined selectors (better perf)
const { activities, language } = yield select(state => ({
   activities: activitiesSelector()(state),
   language: state.language
}));
```

```js
// With 2 calls (more readable)
const activities = yield select(activitiesSelector);
const language = yield select(languageSelector);
```

vs. the bare approach

```js
const { activities, language } = yield select();
```

The benefit here would be ability to quickly adapt to different store shape. The downside is more code to run.

What do you think?

Reply:
@cwhitten I have updated this PR with selector.

Could you help me to review the change? I really want to make the selector architecture right, simple and intuitive to use.

To jumpstart you on the selector architecture:

- [`selectors/language.js`](https://github.com/Microsoft/BotFramework-WebChat/pull/1286/commits/f6737e332f3513df686dd787acd5641c8852e9de#diff-55132b6cc5589cd51c40c50eb106ccdbR1) is the simplest selector, getting the language settings
   - [`sagas/postActivitySaga.js`](https://github.com/Microsoft/BotFramework-WebChat/pull/1286/commits/f6737e332f3513df686dd787acd5641c8852e9de#diff-e5d56925571fab8eb14108fef46d8b23R43) is a sample usage of this selector
- [`sagas/startDictateAfterSpeakActivitySaga.js`](https://github.com/Microsoft/BotFramework-WebChat/pull/1286/commits/f6737e332f3513df686dd787acd5641c8852e9de#diff-95143c2160a015d4a649b9bf088b954fR35) is getting the first activity with ID of `activityID`
   - Note the use of rest operator here to get the first activity out of an array
- [`selectors/activities.js`](https://github.com/Microsoft/BotFramework-WebChat/pull/1286/commits/f6737e332f3513df686dd787acd5641c8852e9de#diff-25ad76dafa72bbef0076b54dd5195217R1) is our most complicated selector, it support a few things
   - Simply selecting all activities
   - Selecting activities using a predicate
   - Selecting activities of a specified type
   - Selecting activities of a specified ID (will return array)
      - Although ID is consider unique and should return a single result, prefer returning array for uniform result
- [`sagas/markAllAsSpokenOnStopSpeakActivitySaga.js`](https://github.com/Microsoft/BotFramework-WebChat/pull/1286/commits/f6737e332f3513df686dd787acd5641c8852e9de#diff-e220fa2015b4cb57f5fe472e3683a34aR19) is a sample usage of combining selector and definition

p.s. this is the last comment for this PR, but feel free to open more.